tox-offline-draft.md

```diff
@@ -5,77 +5,154 @@
-receiver is fully in control of what they want to pull at this point.)
+Users simply check the SENDER's STORAGE NODES in the DHT when they come
+online.  If any messages are available they will download them from the nodes
+holding them and decrypt them with their OM keys. Nodes holding messages
+should then delete them from their store.
 
 ## Spam
 
 As a Proof-of-Work (PoW) hash function is used to insert messages into the
-DHT. The sender must complete a certain amount of computational hashing
-before they can send a message. Messages that do not properly complete the
-hashing function will be dropped by network clients upon receival. Thus is
-makes spamming a user unreliable and costly.
-
+DHT. The SENDER and RECEIVER must complete a certain amount of
+computational hashing before they can send a message. Messages that do not
+properly complete the hashing function will be dropped by network clients
+upon receival. Thus is makes spamming a single user or the network
+unreliable and costly.
+
+The difficulty of the PoW is increased relative to the storage space
+required to storage messages, to ensure a malicious SENDER cannot simply
+fill a STORAGE_NODE's available message space with a single message.
+However, very small messages could game this multiplier, and thus even the
+minimum PoW difficulty must be reasonably large.
+
+Messages should also expire eventually, and thus a higher PoW difficulty
+should be required for a longer message TTL (time to live).
+        
 ## Privacy
 
 Tox does not inherently provide anonymity, however this will not expose any
 more information about the user than using the DHT in Tox already does.
-(TODO: A correlation between who is talking to whom may be possible?).
+
+A middle node can correlate which IPs are talking to each other by seeing who
+requests messages sent by whom, however one could enumerate this simply by
+watching traffic from either the SENDER or RECEIVER so this does not provide
+much new information.
 
 ## Security
 
-Messages are encrypted with the same encryption they normally are, and
-users will not be able to read the content of other user's messages unless
-intended for them.
+Messages are encrypted by the SENDER with two layers, one layer for the
+STORAGE_NODE which encrypts the metadata necessary to see whom the message is
+intended for, and another layer for the RECEIVER that includes the message
+itself. Users will not be able to read the content of other user's messages
+unless intended for them.
 
 # Technical
 
-Technical details below.
-
 ## Packet details
 
 Need a packet type for sending message, sending message TTL updates,
-retreiving messages from inbox.
+retreiving messages from STORAGE_NODEs, sending PoW requests and responses.
+
+## Packet encryption
+
+There are 2 layers of encryption for a message, first the SENDER encrypts the
+message using the agreed/shared key. (TODO: Is this the Tox per-friend key we
+already use for messaging, or is this OM specific?) 
```

Comment:
In most pessimistic case, users cannot be both online for a really long time (especially if we are talking about mobile clients). To have a forward security in that case I was thinking of 2 options:

### Creating multiple keys per friend. 

When both clients are online, they can generate and exchange `K` keys, that will be used by to encrypt offline messages. First key would be used to encrypt first `M` messages, after that it would be dropped. Then second key would be used to store next `M` messages, and so on. `M` can be also increased for next keys.

However I'm not sure how much time does it take to generate shared key, maybe too much resources would be needed for that.

For example, let's say `K` = 10, `M` = 32.
- Alice and Bob are online, they agree on 10 shared keys
- Bob goes offline
- Alice sends 32 messages encrypted by `K1`, `K1` is dropped by Alice
- Alice sends 32 messages encrypted by `K2`, `K2` is dropped by Alice
...
- Alice sends 64 messages encrypted by `K8`
- Alice sends 128 messages encrypted by `K9`
- Alice sends all other messages encrypted by `K10`

At some point both Alice and Bob are online and can generate new K keys.

### Creating new keys using offline messages

Contacts agree on one shared key, then after some amount of messages they exchange new shared key through offline message storage. Not sure how secure this approach is.